lib/galaxy/tools/toolbox/base.py

```diff
@@ -927,53 +927,59 @@ def tool_panel_contents(self, trans, **kwds):
             if elt:
                 yield elt
 
+    def _get_tool_to_dict_from_cache(self, trans, tool_id):
+        """Return tool's to_dict if cached.
+        Note: The cached tool's to_dict is specific to the calls from toolbox.
+        """
+        to_dict_from_cache = None
+        if not trans.user_is_admin():
+            to_dict_from_cache = self._tool_to_dict_cache.get(tool_id, None)
+        else:
+            to_dict_from_cache = self._tool_to_dict_cache_admin.get(tool_id, None)
+        return to_dict_from_cache
```

Comment:
Alternative implementation (not tested) that combines get, eventual calculation, and store:
```python
    def _get_tool_to_dict_from_cache(self, trans, tool):
        """Return tool's to_dict if cached.
        Note: The cached tool's to_dict is specific to the calls from toolbox.
        """
        if not trans.user_is_admin():
            to_dict = self._tool_to_dict_cache.get(tool.id, None)
            if not to_dict:
                to_dict = tool.to_dict(trans, link_details=True)
                self._tool_to_dict_cache[tool.id] = to_dict
        else:
            to_dict = self._tool_to_dict_cache_admin.get(tool.id, None)
            if not to_dict:
                to_dict = tool.to_dict(trans, link_details=True)
                self._tool_to_dict_cache_admin[tool.id] = to_dict
        return to_dict
```